hw/ip/prim/dv/prim_ram_scr/cpp/scramble_model.cc

```diff
@@ -0,0 +1,373 @@
+// width (64-bit).
+static std::vector<uint8_t> scramble_gen_keystream(
+    const std::vector<uint8_t> &addr, uint32_t addr_width,
+    const std::vector<uint8_t> &nonce, const std::vector<uint8_t> &key,
+    uint32_t keystream_width, uint32_t num_half_rounds) {
+  assert(key.size() == (kPrinceWidthByte * 2));
+
+  // Determine how many PRINCE replications are required
+  uint32_t num_prince = (keystream_width + kPrinceWidth - 1) / kPrinceWidth;
+
+  std::vector<uint8_t> keystream;
+
+  for (int i = 0; i < num_prince; ++i) {
+    // Initial vector is data for PRINCE to encrypt. Formed from nonce and data
+    // address
+    std::vector<uint8_t> iv(8, 0);
+
+    for (int j = 0; j < kPrinceWidth; ++j) {
+      if (j < addr_width) {
+        // Bottom addr_width bits of IV are address
+        or_vector_bit(iv, j, read_vector_bit(addr, j));
+      } else {
+        // Other bits are taken from nonce. Each PRINCE instaniations will used
+        // different nonce bits.
+        or_vector_bit(
+            iv, j,
+            read_vector_bit(
+                nonce, (j - addr_width) + i * (kPrinceWidth - addr_width)));
+      }
+    }
+
+    // PRINCE C reference model works on big-endian byte order
+    iv = byte_swap_vector(iv);
+    std::vector<uint8_t> key_be = byte_swap_vector(key);
+
+    // Apply PRINCE to IV to produce keystream
+    std::vector<uint8_t> keystream_block(kPrinceWidthByte);
+    prince_enc_dec(&iv[0], &key_be[0], &keystream_block[0], 0, num_half_rounds,
+                   0);
+
+    // Flip keystream into little endian order and add to keystream vector
+    keystream_block = byte_swap_vector(keystream_block);
+    keystream.insert(keystream.end(), keystream_block.begin(),
+                     keystream_block.end());
+  }
+
+  return keystream;
+}
+
+// Split incoming data into `subst_perm_width` chunks and individually apply the
+// substitution/permutation layer to each
+static std::vector<uint8_t> scramble_subst_perm_full_width(
+    const std::vector<uint8_t> &in, uint32_t bit_width,
+    uint32_t subst_perm_width, bool enc) {
+  assert(in.size() >= ((bit_width + 7) / 8));
+
+  // Determine how many bytes each `subst_perm_width` chunk is and how many
+  // chunks are needed to cover the full `bit_width`.
+  uint32_t subst_perm_bytes = (subst_perm_width + 7) / 8;
+  uint32_t subst_perm_blocks =
+      (bit_width + subst_perm_width - 1) / subst_perm_width;
+
+  std::vector<uint8_t> out(in.size(), 0);
+  std::vector<uint8_t> zero_key(subst_perm_bytes, 0);
+
+  for (int i = 0; i < subst_perm_blocks; ++i) {
+    uint32_t local_subst_perm_width;
+
+    // Where `bit_width` does not evenly divide into `subst_perm_width` the
+    // final chunk is smaller.
+    if (i == subst_perm_blocks - 1) {
+      local_subst_perm_width =
+          bit_width - subst_perm_width * (subst_perm_blocks - 1);
+    } else {
+      local_subst_perm_width = subst_perm_width;
+    }
```

Comment:
I think this might be clearer like this?
```c++
uint32_t bits_so_far = subst_perm_width * i;
uint32_t block_width = std::min(subst_perm_width, bit_width - bits_so_far);
```